### DISCUSSION TITLE
Start Scheduler manually and using DI

### DISCUSSION THREAD

**Discussion by gkarwchan** [quartznet/quartznet#1916] [Q&A]
I am using Quartz inside ASP.NET 6
I want to use the DependencyInjectionJobFactory, and at the same time, I don't want to start the scheduler manually (on response of calling one of the controller POST) and not using the hosted service.
If I add that in startup.cs

```
Services.AddQuartz(q => 
  q.UseMicrosoftDependencyInjectionJobFactory();
);
```

And when I start the scheduler, I am using this code (there is more to the code, but this is simplified to talk about the issue)

```
[HttpPost]
pubic async Task<bool> StartAsync()
{
    // schedulerFactory is injected through DI and resolved as StdSchedulerFactory
   var scheduler = await schedulerFactory.GetScheduler(cancelToken)
   scheduler.start(cancelToken);
}

```
But the job factory won't be the DI one, it will be the default.
How to use the DI job factory with that scenario?

Appreciate your help

**Comment by lahma** [comment]
Can you create a full repro, preferably as GitHub  repository?

**Reply by s3YwCf2ZbfJG4SHAfjQMAjtsf** [reply]
That is exactly the issue I am having.   Delayed start is impacting how the DI comes into play.
https://github.com/quartznet/quartznet/discussions/1914
![image](https://user-images.githubusercontent.com/784740/219770513-f6b14561-2045-4c31-a259-38b87d810239.png)

**Reply by lahma** [reply]
OK so it's now a race between the two of you, who creates a repro repo faster 😉

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
It has been a while   (Hope I did this right)

Project 1
Scheduler started late
`.UseJobFactory<MicrosoftDependencyInjectionJobFactory>()`
Quartz_SchedulerLate_UseJobFactoryByType
https://github.com/s3YwCf2ZbfJG4SHAfjQMAjtsf/QuartzTests

Error
Quartz.SchedulerConfigException
  HResult=0x80131500
  Message=Unable to Instantiate JobFactory: Cannot instantiate type which has no empty constructor (Parameter 'MicrosoftDependencyInjectionJobFactory')
  Source=Quartz
  StackTrace:
   at Quartz.Impl.StdSchedulerFactory.<Instantiate>d__65.MoveNext()
   at Quartz.Impl.StdSchedulerFactory.<GetScheduler>d__71.MoveNext()
   at Quartz.Services.QuartzService.<StartServiceComplexAsync>d__8.MoveNext() in E:\Projects\Quartz\Quartz_SchedulerLate_UseJobFactoryByType\Services\QuartzService.cs:line 47
   at Quartz.Program.<Main>d__0.MoveNext() in E:\Projects\Quartz\Quartz_SchedulerLate_UseJobFactoryByType\Program.cs:line 33

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
ArgumentException: Cannot instantiate type which has no empty constructor (Parameter 'MicrosoftDependencyInjectionJobFactory')

**Reply by lahma** [reply]
I think you need to make the repo public, can't access it.

**Reply by s3YwCf2ZbfJG4SHAfjQMAjtsf** [reply]
:)  Oops sorry, good ready.   Not ready to describe the other folder.

THANKS!

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
Project 2
Scheduler started late
Quartz_SchedulerLate_JobFactoryViaProgram
https://github.com/s3YwCf2ZbfJG4SHAfjQMAjtsf/QuartzTests

Sorry, I am using Console logging, but I am not seeing the log entry indicating my job started and it won't debug into the single line I have in TestJob.   As shown by the screen shot, this example is now similar to this new post and the screen shot indicates that isn't using the correct JobFactory.


![image](https://user-images.githubusercontent.com/784740/219791045-54341dd5-51d3-4fe5-a03a-3bafa8d67289.png)

**Reply by lahma** [reply]
This problem is fixed by adding the job factory initialization line like I've tried to say in earlier discussion thread. This is the way if you don't want to use the direct DI integration and want to use `SchedulerBuilder` directly (which doesn't know about any DI).

```
// add this before start, serviceProvider is from app.Services or some other initialization
_scheduler.JobFactory = new MicrosoftDependencyInjectionJobFactory(serviceProvider, Options.Create(new QuartzOptions()));

await _scheduler.Start();
```

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
THANK YOU!

Complete solution with delayed start (aka after app.Run):

```
        public static IServiceProvider siteService;
// Main...
            // This confirms Quartz is being started after "app.Run".
            // However it could be started by a UI/controller later in the app's lifecycle.
            var task = Task.Run(async () =>
            {
                Thread.Sleep(1000);
                await QuartzService.StartServiceComplexAsync();
            });
            // This will not always be here. 
            //await QuartzService.StartServiceComplexAsync();

            siteService = app.Services;
```


// add this before start, serviceProvider is from app.Services or some other initialization, using Program.siteService for IServiceProvider.
_scheduler.JobFactory = new MicrosoftDependencyInjectionJobFactory(**Program.siteService,** Options.Create(new QuartzOptions()));

await _scheduler.Start();

Debugger showing JobFactory updated:
![image](https://user-images.githubusercontent.com/784740/219852806-01c97753-08a4-4402-8eea-4304c373407b.png)

Debugger showing that TestJob has ILogger available:
![image](https://user-images.githubusercontent.com/784740/219852838-6fa1ce7f-806e-4087-ad93-dd8c612675c0.png)
